samples/js/photos/src/photo-grid-item.js

```diff
@@ -43,53 +50,96 @@ export default class PhotoGridItem extends React.Component<void, Props, State> {
   constructor(props: Props) {
     super(props);
     this.state = {
-      size: props.gridSize,
-      sizeIsBest: false,
-      url: props.url,
+      blob: this.getBlob(props),
+      tx: false,
     };
     this._parentTop = 0;
     this._parentLeft = 0;
-    this._shouldTransition = true;
+    this.load(this.state.blob);
   }
 
-  componentWillUpdate() {
+  componentWillReceiveProps(nextProps) {
+    const nextBlob = this.getBlob(nextProps);
+    if (nextBlob == this.state.blob) { 
+      return;
+    }
+    this.setState({
+      blob: nextBlob,
+    });
+    this.load(nextBlob);
+  }
+
+  componentWillUpdate(nextProps, nextState) {
+    if (this.state.url != nextState.url) {
+      this.releaseBlobURL();
+    }
+
     const rect = ReactDOM.findDOMNode(this).parentElement.getBoundingClientRect();
     this._parentTop = rect.top;
     this._parentLeft = rect.left;
   }
 
+  getBlob(props: Props): NomsBlob {
+    let [w, h] = props.fullscreen ?
+      [props.viewport.clientWidth, props.viewport.clientHeight] :
+      [props.gridSize.width, props.gridSize.height];
+    const [_, blob] = props.photo.getBestSize(w, h);
+    return blob;
+  }
+
+  async load(blob: NomsBlob) {
+    const b = await blob.targetValue(this.props.db);
+    const r = b.getReader();
+    const parts = [];
+    while (true) {
+      const n = await r.read();
+
+      // might have changed
+      if (this.state.blob != blob) {
+        return;
+      }
+
+      if (n.value) {
```

Comment:
It's an invariant that there is a value if !done, Flow should even know that: https://github.com/attic-labs/noms/blob/master/js/noms/src/async-iterator.js#L11

So it would be better to structure this like
```javascript
while (true)
  const n = await.read();
  if (n.done) break;
  if (this.state.blob !== blob) return;
  parts.push(n.value);
}
```
modulo formatting.